File: robot_arm_example/py_subscriber/arm_accel_sub.py
#!/usr/bin/env python3
"""Robot arm acceleration subscriber with motion detection."""

import math
import time
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
from geometry_msgs.msg import Accel

MOTION_THRESHOLD = 0.5  # g — acceleration magnitude above which to trigger
COOLDOWN_SECONDS = 2.0  # minimum time between motion events


class ArmAccelSubscriber(Node):
    """Subscribes to /adxl355/accel and detects motion events."""

    def __init__(self):
        super().__init__('arm_accel_subscriber')

        qos = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10)

        self._sub = self.create_subscription(
            Accel, '/adxl355/accel', self._callback, qos)

        self._state = 'IDLE'
        self._last_event_time = 0.0
        self._msg_count = 0
        self._last_log_time = time.time()

        self.get_logger().info(
            f'Arm subscriber started. Threshold: {MOTION_THRESHOLD}g | '
            f'Cooldown: {COOLDOWN_SECONDS}s')

    def _callback(self, msg: Accel):
        self._msg_count += 1

        ax = msg.linear.x
        ay = msg.linear.y
        az = msg.linear.z
        magnitude = math.sqrt(ax * ax + ay * ay + az * az)

        now = time.time()

        if self._state == 'IDLE':
            if magnitude > MOTION_THRESHOLD:
                self._state = 'MOVING'
                self._last_event_time = now
                self.get_logger().info(
                    f'MOTION DETECTED | mag={magnitude:.3f}g | '
                    f'ax={ax:.3f} ay={ay:.3f} az={az:.3f}')
                self._on_motion_start(magnitude, ax, ay, az)

        elif self._state == 'MOVING':
            if magnitude <= MOTION_THRESHOLD:
                if now - self._last_event_time > COOLDOWN_SECONDS:
                    self._state = 'IDLE'
                    self.get_logger().info(
                        f'Motion ended. Returning to IDLE.')
                    self._on_motion_end()

        # Periodic rate logging
        if now - self._last_log_time >= 5.0:
            rate = self._msg_count / (now - self._last_log_time)
            self.get_logger().info(
                f'Status: {self._state} | Rate: {rate:.1f} Hz')
            self._msg_count = 0
            self._last_log_time = now

    def _on_motion_start(self, mag: float, ax: float, ay: float, az: float):
        """Called when vibration exceeds threshold.
        Replace with actual robot arm control logic."""
        # Example: send command to robot arm
        # self._arm_client.send_command(
        #     f'compensate_vibration {ax:.4f} {ay:.4f} {az:.4f}')
        pass

    def _on_motion_end(self):
        """Called when vibration drops below threshold for cooldown period.
        Replace with actual robot arm control logic."""
        # Example: return arm to normal position
        # self._arm_client.send_command('return_to_idle')
        pass


def main():
    rclpy.init()
    node = ArmAccelSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: robot_arm_example/cpp_subscriber/src/arm_accel_sub.cpp
/**
 * Robot arm acceleration subscriber (C++).
 * Subscribes to /adxl355/accel and detects motion events.
 */

#include <chrono>
#include <cmath>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/accel.hpp"

using namespace std::chrono_literals;

class ArmAccelSubscriber : public rclcpp::Node {
public:
  ArmAccelSubscriber()
  : Node("arm_accel_subscriber"), state_(State::IDLE) {

    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.best_effort();

    subscription_ = this->create_subscription<geometry_msgs::msg::Accel>(
      "/adxl355/accel", qos,
      std::bind(&ArmAccelSubscriber::callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(),
      "Arm subscriber started. Threshold: %.1fg | Cooldown: %.1fs",
      MOTION_THRESHOLD, COOLDOWN_SECONDS);
  }

private:
  enum class State { IDLE, MOVING };

  static constexpr double MOTION_THRESHOLD = 0.5;  // g
  static constexpr double COOLDOWN_SECONDS = 2.0;

  rclcpp::Subscription<geometry_msgs::msg::Accel>::SharedPtr subscription_;
  State state_;
  rclcpp::Time last_event_time_{0, 0, RCL_ROS_TIME};

  void callback(const geometry_msgs::msg::Accel::SharedPtr msg) {
    double ax = msg->linear.x;
    double ay = msg->linear.y;
    double az = msg->linear.z;
    double magnitude = std::sqrt(ax * ax + ay * ay + az * az);

    auto now = this->now();

    if (state_ == State::IDLE) {
      if (magnitude > MOTION_THRESHOLD) {
        state_ = State::MOVING;
        last_event_time_ = now;
        RCLCPP_INFO(this->get_logger(),
          "MOTION DETECTED | mag=%.3fg | ax=%.3f ay=%.3f az=%.3f",
          magnitude, ax, ay, az);
        on_motion_start(magnitude, ax, ay, az);
      }
    } else if (state_ == State::MOVING) {
      if (magnitude <= MOTION_THRESHOLD) {
        double elapsed = (now - last_event_time_).seconds();
        if (elapsed > COOLDOWN_SECONDS) {
          state_ = State::IDLE;
          RCLCPP_INFO(this->get_logger(),
            "Motion ended. Returning to IDLE.");
          on_motion_end();
        }
      }
    }
  }

  void on_motion_start(double mag, double ax, double ay, double az) {
    // TODO: Replace with actual robot arm control logic.
    // Example:
    //   arm_client_->send_command("compensate_vibration",
    //                             ax, ay, az);
  }

  void on_motion_end() {
    // TODO: Replace with actual robot arm control logic.
    // Example:
    //   arm_client_->send_command("return_to_idle");
  }
};

int main(int argc, char * argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ArmAccelSubscriber>());
  rclcpp::shutdown();
  return 0;
}
